### PR TITLE
[wip] Call configure_module before freeze_before_training

### LINKED ISSUE #19658: `configure_model` is incompatible with the `BaseFinetuning` behavior when fitting
### Bug description

Based on the current callback orders, The `Finetuning` class will always be incompatible with any `LightningModule` that utilize `configure_model` method. The current callback sequence is `Callback.setup` -> `LightningModule.configure_model` -> `LightningModule.configure_optimizers` -> `Callback.on_fit_start`. However, The `BaseFinetuning` calls `freeze_before_training` at `setup`, where modules inside the `configure_model` has not been instantiated yet.

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug

```python
from lightning import LightningModule
import torch
from torchvision import models
class MyModel(LightningModule):
    def configure_model(self):
        self.backbone = models.resnet18()
    def configure_optimizers(self):
        return torch.optim.SGD(lr=1e-3)
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by chualanagit**
## What does this PR do?

Fixes #19658

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by lantiga** [comment]
Thank you @chualanagit 

We need to make sure configure_model is not called twice:
https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/trainer/trainer.py#L945

is that the case?

**Comment by chualanagit** [comment]
> Thank you @chualanagit
> 
> We need to make sure configure_model is not called twice: https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/trainer/trainer.py#L945
> 
> is that the case?

addressed that in the test case, please review! cc @lantiga

**Comment by chualanagit** [comment]
> configure_model

I double checked all the tests that failed, seems like all the failures are identical, failing on assertion that `setup()` needs to be called before `configure_model()`. I wonder if the assertions themselves need to be changed? There seem to be no other errors caught by any other test cases can that indicates this change leads to any logical failures in the library. Pushing a change to modify the assertion order in the `test_hooks.py` files and see if any other tests are failing.

**Comment by chualanagit** [comment]
> f the assertions themselves need to be changed? There seem to be no other errors caught by any other test cases can that indicates this change leads to any logical failures in the library. Pushing a change to modify the assertion order in the `test_hooks.py` files and see if any other tests are failing

seems like no other errors are raised, wondering if anyone knows of any reason why `configure_model` hook should not be called before `setup` hook?

**Comment by chualanagit** [comment]
@lantiga quick question, how reliable in general are the CI tests? I have definitely seen transient errors that show up on a specific commit and disappear in another before. Also seen before that a PR was merged when some CI tests are still failing. Would love to learn more for future contributions, thanks!

**Comment by lantiga** [comment]
hey @chualanagit we made improvements to CI stability last week, but this is a complex project with lots of tests on multiple backends, so things may run into hiccups (help very welcome there :-) )

**Comment by lantiga** [comment]
Hey @chualanagit just so you know, we cleaned up CI quite a bit, things are running much smoother now.

**Review by Borda — APPROVED** [review]

**Comment by GdoongMathew** [comment]
I think this change could break some existing user behavior, especially if their implementation relies on the fact that modules defined in `configure_model` aren’t instantiated inside the `LightningModule` before `setup` hook is called.

**Comment by deependujha** [comment]
Hi @GdoongMathew, One way to fix this for BaseFinetuning while keeping backward-compatibility is to introduce a new hook. Two options:

1. **`pre_setup`**:

   `call_pre_setup_hook -> call_configure_model -> call_setup_hook`
   Users who relied on `setup` running first would need to move their code to `pre_setup`.

2. **`post_setup` (non-breaking)**:

   `call_setup_hook -> call_configure_model -> call_post_setup_hook`
   BaseFinetuning could use `post_setup`, and existing users remain unaffected.

Both approaches make finetuning compatible with modules that rely on `configure_model` (FSDP, etc.).

**Comment by deependujha** [comment]
As suggested by @SkafteNicki , it makes more sense to move BaseFinetuning's `setup()` code to `on_fit_start()`.

**Comment by deependujha** [comment]
Update:

`on_fit_start` won’t work here. For BaseFinetuning, `freeze_before_training` must be called before `configure_optimizers`. 

https://github.com/Lightning-AI/pytorch-lightning/blob/79dc82c66cc38523abaf71a42fc5ebc8e06fb015/src/lightning/pytorch/callbacks/finetuning.py#L50-L51

Current order:
```
setup -> configure_model -> configure_optimizers (via strategy) -> on_fit_start
```

We need a hook between `configure_model` and `configure_optimizers`.

 So we can introduce `on_before_optimizer_setup`:
 
```
setup -> configure_model -> on_before_optimizer_setup -> configure_optimizers (via strategy) -> on_fit_start
```

This ordering works perfectly and keeps BaseFinetuning behavior consistent with its docs.

---

Introducing the hook in a separate PR: #21272
